### ISSUE TITLE
Wrong return type inference when `cls` is annotated as `Type[T]` and `T` is bound to protocol

### ISSUE THREAD

**Issue by jack1142** [microsoft/pyright#2187]
**Describe the bug**
Annotating `cls` (in classmethod) as `Type[T]` when `T` is bound to a Protocol breaks inference for the return type of that method.

**To Reproduce**
1. Copy code from below
2. Run it, you should see:
```
<__main__.Sample object at 0x00000162327A7A90>
```
3. Type check it with pyright, as part of the output you should see:
```
sample.py:22:17 - info: Type of "Sample.test()" is "Type[Sample]"
sample.py:23:17 - info: Type of "Sample().test()" is "Type[Sample]"
```

**Expected behavior**
I expected pyright to infer type of `Sample.test()`/`Sample().test()` correctly:
```
sample.py:22:17 - info: Type of "Sample.test()" is "Sample"
sample.py:23:17 - info: Type of "Sample().test()" is "Sample"
```

**Screenshots or Code**
```python
from typing import Any, Optional, Protocol, Union, Type, TypeVar, TYPE_CHECKING, overload


class ClsProtocol(Protocol):
    def __init__(self):
        ...


T = TypeVar("T", bound="ClsProtocol")


class Sample:
    @classmethod
    def test(cls: Type[T]) -> T:
        return cls()


print(Sample.test())
print(Sample().test())

if TYPE_CHECKING:
    reveal_type(Sample.test())
    reveal_type(Sample().test())
```

**VS Code extension or command-line**
Running pyright 1.1.161 as a command-line tool. Replicable on the build from the current state of `main` branch as well.

**Additional context**
The code I originally found this in is pretty much as what you can see in #2177 and I wanted to report it there initially but it seems it is reproducible with a simpler classmethod as well.

mypy infers the type correctly:
```
sample.py:22: note: Revealed type is "descriptor_fun.Sample*"
sample.py:23: note: Revealed type is "descriptor_fun.Sample*"
```

Example test case from mypy:
https://github.com/python/mypy/blob/644496828841035d097d7afd80b96081e0ea5c7f/test-data/unit/check-selftype.test#L784-L803

**Comment by jack1142** [comment]
As an additional example, here's an equivalent reproduction code for metaclasses (only reproducible on `main` due to #2177):
```py
from typing import Any, Optional, Protocol, Union, Type, TypeVar, TYPE_CHECKING, overload



class ClsProtocol(Protocol):
    def __init__(self):
        ...


T = TypeVar("T", bound=ClsProtocol)


class _IntDescriptorMeta(type):
    @overload
    def __get__(self: Type[T], instance: None, owner: Any) -> Type[T]: ...

    @overload
    def __get__(self: Type[T], instance: object, owner: Any) -> T: ...

    def __get__(self: Type[T], instance: Optional[object], owner: Any) -> Union[T, Type[T]]:
        if instance is None:
            return self
        return self()


class IntDescriptorClass(metaclass=_IntDescriptorMeta):
    ...


class X:
    number_cls = IntDescriptorClass


print(X.number_cls)
print(X().number_cls)

if TYPE_CHECKING:
    reveal_type(X.number_cls)
    reveal_type(X().number_cls)
```

**Comment by erictraut** [comment]
Thanks for the bug report. This will be fixed in the next release.

FWIW, mypy does not produce the correct results on the second sample.

**Comment by erictraut** [comment]
This is included in Pyright 1.1.163, which I just published. It will also be in the next release of Pylance.
